imap_processing/swe/l1a/swe_science.py

```diff
@@ -0,0 +1,168 @@
+        5: {"base": 256, "step_size": 16},
+        6: {"base": 512, "step_size": 16},
+        7: {"base": 768, "step_size": 16},
+        8: {"base": 1024, "step_size": 32},
+        9: {"base": 1536, "step_size": 32},
+        10: {"base": 2048, "step_size": 64},
+        11: {"base": 3072, "step_size": 128},
+        12: {"base": 5120, "step_size": 256},
+        13: {"base": 9216, "step_size": 512},
+        14: {"base": 17408, "step_size": 1024},
+        15: {"base": 33792, "step_size": 2048},
+    }
+
+    # uncompression formula from SWE algorithm document CN102D-D0001 and page 16.
+    # N = base[index] + multi * step_size[index] + (step_size[index] - 1) / 2
+    # NOTE: for (step_size[index] - 1) / 2, we only keep the whole number part of
+    # the quotient
+
+    return (
+        uncompress_table[index]["base"]
+        + (multi * uncompress_table[index]["step_size"])
+        + ((uncompress_table[index]["step_size"] - 1) // 2)
+    )
+
+
+def add_metadata_to_array(data_packet, metadata_arrays):
+    """This function add metadata to metadata_arrays.
+
+    Parameters
+    ----------
+    data_packet : space_packet_parser.parser.Packet
+        SWE data packet
+    metadata_arrays : dict
+        metadata arrays
+    """
+    for key, value in data_packet.header.items():
+        metadata_arrays.setdefault(key, []).append(value.raw_value)
+
+    for key, value in data_packet.data.items():
+        if key != "SCIENCE_DATA":
+            metadata_arrays.setdefault(key, []).append(value.raw_value)
+
+
+def swe_science(decom_data):
+    """SWE L1A algorithm steps:
+        - Read data from each SWE packet file
+        - Uncompress counts data
+        - Store metadata fields and data in DataArray of xarray
+        - Save data to dataset
+
+    In each packet, SWE collects data for 15 seconds. In each second,
+    it collect data for 12 energy steps and at each energy step,
+    it collects 7 data from each 7 CEMs.
+
+    Each L1A data from each packet will have this shape: 15 rows, 12 columns,
+    and each cell in 15 x 12 table contains 7 element array.
+    These dimension maps to this:
+        15 rows --> 15 seconds
+        12 column --> 12 energy steps in each second
+        7 element --> 7 CEMs counts
+
+    In L1A, we don't do anything besides read raw data, uncompress counts data and
+    store data in 15 x 12 x 7 array.
+
+    SWE want to keep all value as it is in L1A. Post L1A, we group data into full cycle
+    and convert raw data to engineering data as needed.
+
+    Parameters
+    ----------
+    packet_file : str
+        packet file path
+
+    Returns
+    -------
+    xarray.Dataset
+        xarray dataset with data.
+    """
+    science_array = []
+
+    metadata_arrays = collections.defaultdict(list)
+
+    # We know we can only have 8 bit numbers input, so iterate over all
+    # possibilities once up front
+    decompression_table = np.array([uncompress_counts(i) for i in range(256)])
```

Comment:
If the table is static and doesn't change, does it make sense to just have it live in a csv file somewhere?  I guess it's less flexible in case the table needs to be changed.  But it might help people understand what is going on a little better.